### PR TITLE
Correct password option docs and add related tests

### PR DESCRIPTION
**Pull request by fidgetingbits**
## Description of changes

This attempts to clarify the true behavior of password option overrides, as the documentation seems to be wrong (and appears to have been for at least 10 years). These docs also clarify systemd-sysusers quirks that I coincidentally noticed someone happened to file recently [here](https://github.com/NixOS/nixpkgs/issues/307159) as well.

The relevant option docs have been updated, but more importantly I created two new tests to try to show the true behavior of the overrides in their current form. This way if anything ever changes on this front it will be caught and docs can be updated accordingly.

If you want to run the tests I added, you will have to change the maintainer for now as my being added to maintainers is pending [this PR](https://github.com/NixOS/nixpkgs/pull/296562) being merged.

I've not used Nix for very long, so would appreciate someone who has dealt with this code in the past to validate what I've found. @nikstur and @NeQuissimus, perhaps you could both take a look as I've repurposed some of your tests that touch the same areas, so I suspect you'll both be familiar. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by fidgetingbits** [comment]
@Ma27 any thoughts on this PR or know who should review it? I just noticed today you recently made some recent changes that lead to a merge conflict (which I have yet to fix), so just pinging to see if you have want to take a look or have an idea who to ping to get this reviewed.

**Comment by Ma27** [comment]
Last times I touched this part, it was kinda hard to find anybody who felt responsible.
Given that this seems to be a correctness fix mostly, I'll try to take a look soonish.

**Comment by fidgetingbits** [comment]
Merge conflicts should be fixed now

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [inline nixos/modules/config/users-groups.nix:361]: Hmm after reading through the code it seems to me as if all non-initial password options would override this.

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [inline nixos/modules/config/users-groups.nix:73]: I think you can cross-reference this option with `[](#opt-systemd.sysusers.enable)`.

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [inline nixos/modules/config/users-groups.nix:976]: What about the other two options?

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [inline nixos/modules/config/users-groups.nix:969]: The trailing comma should go here: there's an `&` the next char.

Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix]: TIL

File: nixos/modules/config/users-groups.nix
Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix]: The other two options aren't supported when using `systemd.sysusers.enable`. I changed that to use some other (hopefully more clear) wording from elsewhere. So should be addressed, but let me know if you still think it's too unclear.

File: nixos/modules/config/users-groups.nix
Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix]: I might not understand what you mean, but I think the test on line 35 of nixos/tests/password-option-override-ordering.nix that I added shows that initialPassword actually overrides hashedPassword (a non-initial password option), which conflicts with what you are saying?

**Comment by fidgetingbits** [comment]
Just a note that to run the tests I added you can run the following from nixpkgs root folder: `nix-build . -A nixosTests.password-option-override-ordering` and `nix-build . -A nixosTests.systemd-sysusers-password-option-override-ordering`. 

Looks like the latter is failing now due to a new assertion about user types, so I will have to investigate.

**Comment by fidgetingbits** [comment]
Looks like there are more aggressive asserts that helps with some of the systemd-sysusers stuff, so I removed a couple of assertions from that test that no longer will run. Also changed the user type and added groups to ensure other new assertions pass so the test actually runs.

I've left the changes as separate commit now for sake of reference, but can squash them before merge.

File: nixos/tests/password-option-override-ordering.nix
Comment (Ma27) [inline nixos/tests/password-option-override-ordering.nix:129]: This seems a little brittle as it only depends on the timing. After reading through all the checks: is that really necessary?

File: nixos/tests/password-option-override-ordering.nix
Comment (Ma27) [inline nixos/tests/password-option-override-ordering.nix:115]: We could match the output of the hashed password from `getent` against the password set for greg here I think.

Also, it's a little weird that you return `passwd` rather than shadow above?

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [inline nixos/modules/config/users-groups.nix:62]: Line 57 of the password-option-* test says `hashedPassword -> ... -> password`, so the override is the other way round here, right?

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [reply nixos/modules/config/users-groups.nix]: Only for `mutableUsers = false`; though, see https://github.com/NixOS/nixpkgs/blob/400e3ecf59205c6197e093e571e6518a669c4d55/nixos/modules/config/users-groups.nix#L417-L421

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [reply nixos/modules/config/users-groups.nix:62]: Also given that this also depends on `mutableUsers` somehow, I guess this is also worth mentioning.

**Review by Ma27 — CHANGES_REQUESTED** [review]
Oof thank you for investing time into this! Using NixOS for almost 8 years now and I had no idea _how_ much of a mess that is...

**Comment by Ma27** [comment]
Also asked in the nixos-systemd Matrix channel for more reviewers for the systemd-sysusers part.

**Comment by Ma27** [comment]
I'd really like to see this in 24.11 when it's good to go: this isn't a behavioral fix, but clarifies the current behavior in both warnings and the manual (and locks it down with a test).

File: nixos/tests/password-option-override-ordering.nix
Comment (Ma27) [reply nixos/tests/password-option-override-ordering.nix]: > Can only check exact match when providing the hash yourself.

Sure, but there are tools that you can use to verify the hash against the plaintext password.

> By return you mean what is printed ya? I switched them all to print getent shadow. Makes sense.

Correct.

File: nixos/tests/password-option-override-ordering.nix
Comment (Ma27) [reply nixos/tests/password-option-override-ordering.nix]: Oof.

But well, shadow is part of the tested set, so it seems to be surprisingly stable :shrug:

File: nixos/tests/password-option-mutable-override-ordering.nix
Comment (Ma27) [inline nixos/tests/password-option-mutable-override-ordering.nix:22]: Wouldn't it make sense to merge this with the immutable test, i.e. have two nodes in there and deduplicate the config and assertions as much as possible?

File: nixos/tests/password-option-override-ordering.nix
Comment (fidgetingbits) [reply nixos/tests/password-option-override-ordering.nix]: > > Can only check exact match when providing the hash yourself.
> 
> Sure, but there are tools that you can use to verify the hash against the plaintext password.

Ok. I'll add that

File: nixos/tests/password-option-mutable-override-ordering.nix
Comment (fidgetingbits) [reply nixos/tests/password-option-mutable-override-ordering.nix]: Done, also deduplicated the login test part.

File: nixos/tests/password-option-override-ordering.nix
Comment (fidgetingbits) [reply nixos/tests/password-option-override-ordering.nix]: Actually I spent some time digging around / testing while I was deduplicating the code and it turns out it's easy to drop the sleep. I'll plan to send some PRs to drop the sleeps from `shadow.nix` and any others I find to speed up tests in the future.

**Comment by fidgetingbits** [comment]
I still need to add the clearer wording for the greg case with mutableUsers difference, but otherwise all other feedback should be addressed now.

File: nixos/modules/config/users-groups.nix
Comment (EmergentMind) [inline nixos/modules/config/users-groups.nix:339]: What is the exception?
Perhaps editing to something like: "... is true, in which case only `initialPassword`, `initialHashedPassword`, and `hashedPasswordFile` options will be used."

This also applies to Note one line 360

**Review by EmergentMind — CHANGES_REQUESTED** [review]
Some suggestions to add clarity and improve some grammar. 

I also reviewed the various cases of precedence depending on the values of mutable and/or sysUsers. All the option descriptions and related tests appear to be consistent.

**Comment by fidgetingbits** [comment]
I took a stab at rewording the whole thing based on EmergentMind's feedback and because I found myself getting confused reading it even though I'm working on it. That allowed me to duplicate some text from other options, which makes `passwordDescription` the single source of truth, which will be nice for future tweaks. 

One part I didn't include which I considered putting at the end of `passwordDescription` is:

```
    If [](#opt-systemd.sysusers.enable) is `true`, then the order of password is:
    `initialPassword` -> `initialHashedPassword` -> `hashedPasswordFile`
```

Since technically that order still matches what is shown earlier (just with password, hashedPassword, hashedPasswordFile omitted), I wasn't sure if I should explicitly include it or not. Happy to add it if others thing it's worth.

File: nixos/modules/config/users-groups.nix
Comment (fidgetingbits) [inline nixos/modules/config/users-groups.nix:984]: I guess this is not true because of the mutableUsers true/false difference, so the order of precedence isn't necessary as shown above. Would be nice to refer to a non-duplicated source of truth here as well, so maybe I'll add a separate variable with the order for mutable false/true text and then include that both in passwordDescription and here.. 

I also wonder if the order of precedence difference between mutableUsers true/false is intentional or accidental. If the former, I wonder who would know, and if the latter, I think it should just be changed for a future release so the orders are consistent?

**Comment by fidgetingbits** [comment]
Good for another review pass I think. In order to fix the assert warning being wrong, and make it harder to mess up the override order text in various places, I tried to deduplicate it all and use some variable expansion to help generate it in various places.

**Review by EmergentMind — COMMENTED** [review]
The rewording and de-duplication is a big improvement for clarity. Nice work.

File: nixos/modules/config/users-groups.nix
Comment (Ma27) [inline nixos/modules/config/users-groups.nix:58]: Isn't there an arrow missing? Same for the ordering below.

**Review by Ma27 — COMMENTED** [review]
Code LGTM otherwise.
Please reword your commit messages to follow the contribution guidelines.

File: nixos/modules/config/users-groups.nix
Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix]: Thanks for catching that!

File: nixos/modules/config/users-groups.nix
Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix:68]: I've rebased with new commits following the contribution guideline and at the same time I _think_ I addressed this suggestion, but please take a look at confirm it's what you wanted. 

I will look into how to actually generate the new docs from nixpkgs, so I can confirm things are formatted sanely.

Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix:68]: Hah.. there is indeed a rendering error I'll fix, but figured I'd note it down here anyway. Built with `nix-build nixos/release.nix -A manual.x86_64-linux` and checked the `options.html`:

> If the option users.mutableUsers is ``, then the order of precedence is as shown below, where values on the left are overridden by values on the right: initialHashedPassword -> hashedPassword -> initialPassword -> password -> hashedPasswordFile

> If the option users.mutableUsers is 1, then the order of precedence is as shown below, where values on the left are overridden by values on the right: initialHashedPassword -> initialPassword -> hashedPassword -> password -> hashedPasswordFile


Is because:
```
nix-repl> builtins.toString true
"1"

nix-repl> builtins.toString false
""
```

Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix:68]: The rendering issue is fixed now

Comment (Ma27) [reply nixos/modules/config/users-groups.nix:68]: > Is because:

Heh, didn't even notice it during review.

Fwiw, this is (IIRC) because Nix does `toString` on all derivation attributes before putting those into the env to make them accessible for the builder (almost always sh/bash) and it kinda makes sense to have an empty string for false in such a language. But yeah, it's a surprise the first time for everybody ;-)

Comment (fidgetingbits) [reply nixos/modules/config/users-groups.nix:68]: Thanks for the insight! It seemed a bit cursed at first, but in the context of generating shellscript output it does indeed make a lot of sense.

**Review by Ma27 — APPROVED** [review]
Pushed two small fixes (one for the whitespace, one for an assertion in the sysusers test - details are in the messages). LGTM now.

**Comment by fidgetingbits** [comment]
Thanks a lot for all your help working through this!

**Pull request merged**
